bblocks/import_tools/imf.py

```diff
@@ -0,0 +1,175 @@
+        A pandas dataframe with SDR data
+    """
+
+    df = pd.read_csv(url, delimiter="/t", engine="python").loc[3:]
+    df = df["SDR Allocations and Holdings"].str.split("\t", expand=True)
+    df.columns = ["member", "holdings", "allocations"]
+
+    return (df.melt(id_vars='member', value_vars=["holdings", "allocations"])
+            .pipe(clean_numeric_series, series_columns="value")
+            .rename(columns={"variable": "indicator"})
+            .reset_index(drop=True)
+            )
+
+
+def _check_indicators(indicators: str | list) -> list:
+    """ """
+
+    accepted_indicators = ['allocations', 'holdings']
+    if isinstance(indicators, str):
+        if indicators not in accepted_indicators:
+            raise ValueError(f"{indicators} is not a valid indicator")
+        return [indicators]
+    elif isinstance(indicators, list):
+        for indicator in indicators:
+            if indicator not in accepted_indicators:
+                raise ValueError(f"{indicator} is not a valid indicator")
+        return indicators
+    elif indicators is None:
+        return ['allocations', 'holdings']
+
+
+class SDR(ImportData):
+    """ """
+
+    indicators: list = field(default_factory=list)
+
+    def load_indicator(self, indicators: Optional | str = None) -> ImportData:
+        """ """
+
+        # make sure indicators are valid
+        indicator_list = _check_indicators(indicators)
+
+        if not os.path.exists(f"{PATHS.imported_data}/{self.file_name}") or self.update_data:
+            self.update()
+
+        df = pd.read_csv(f"{PATHS.imported_data}/{self.file_name}")
+        self.data = df.loc[df['indicator'].isin(indicator_list)].reset_index(drop=True)
+        self.indicators = indicator_list
+
+        return self
+
+    def update(self) -> ImportData:
+        """ """
+
+        base = 'https://www.imf.org/external/np/fin/tad/'
+        # check latest year
+        years = _parse_sdr_links(url='https://www.imf.org/external/np/fin/tad/extsdr1.aspx', concat_url=base)
+        latest_year_link = list(years.values())[0]
+
+        # check latest date
+        dates = _parse_sdr_links(latest_year_link, base)
+        latest_date_link = list(dates.values())[0]
+        latest_date = datetime.strptime(list(dates.keys())[0], "%B %d, %Y").strftime(
+            "%d %B %Y")  # assign latest date
+
+        # find tsv file
+        tsv_link = _get_tsv_url(latest_date_link)
+
+        # read tsv file
+        df = _read_sdr_tsv(tsv_link)
+
+        (df.assign(date=latest_date)
+         .to_csv(f"{PATHS.imported_data}/{self.file_name}", index=False))
+
+        return self
+
+    def get_data(self,
+                 indicators: Optional | str = None,
+                 members: Optional | str | list = None
+                 ) -> pd.DataFrame:
+        """ """
+
+        df = self.data
+
+        indicator_list = _check_indicators(indicators)
+        df = df.loc[df['indicator'].isin(indicator_list)].reset_index(drop=True)
+
```

Comment:
See my comments about this check above. As before indicators is Optional or str.